presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PickTableLayout.java

```diff
@@ -90,14 +100,55 @@ private static boolean shouldRewriteTableLayout(TableScanNode source)
         @Override
         public Result apply(FilterNode filterNode, Captures captures, Context context)
         {
+            TableScanNode tableScan = captures.get(TABLE_SCAN);
+
             TableLayoutRewriter tableLayoutRewriter = new TableLayoutRewriter(metadata, context.getSession(), context.getSymbolAllocator(), context.getIdAllocator());
             PlanNode rewrittenTableScan = tableLayoutRewriter.planTableScan(captures.get(TABLE_SCAN), filterNode.getPredicate());
 
-            if (rewrittenTableScan instanceof TableScanNode || rewrittenTableScan instanceof ValuesNode || (((FilterNode) rewrittenTableScan).getPredicate() != filterNode.getPredicate())) {
-                return Result.ofPlanNode(rewrittenTableScan);
+            if (arePlansEquivalent(filterNode, tableScan, rewrittenTableScan, context)) {
+                return Result.empty();
+            }
+
+            return Result.ofPlanNode(rewrittenTableScan);
+        }
+
+        public boolean arePlansEquivalent(FilterNode filter, TableScanNode tableScan, PlanNode rewritten, Context context)
```

Comment:
- `private`
- `s/filter/originalFilter/`, `s/tableScan/originalTableScan`

Reply:
I removed that method